### PR TITLE
[frontend/backend] (PersoSpace): Transfer Perso Space (#770)

### PR DESCRIPTION
**Pull request by hervyt**
# Context
> Ability for a user to transfer personal space

## How to test
- Connect as user 1 
- Subscribe to a service 
- Request an email transfer on user2
- Go check user2 email 
- CLick on the "Transfer" button 
- User 1 should have lost his subscriptions and user2 should have earn these subscriptions. (view full vidéo démo on the notion page)

## Related issues

- Related to #770

# Checklist

## Quality

- [X] I consider this work complete and ready for review
- [X] I tested all features and edge cases
- [X] I refactored code where necessary to improve quality
- [X] I made sure my code meets development guidelines
- [X] I performed a self-review of my code

## Testing

- [X] I wrote test cases for the relevant use cases (unit, integration or e2e tests)
- [ ] (Bug fixes only) I added tests that reproduce and verify the fix

### What tests have been made

- [ ] Unit tests
- [ ] Integration tests
- [ ] E2E tests
- [X] Local manual tests

### CODE REVIEW

File: apps/portal-api/src/modules/common/user-organization.domain.ts [hunk 1]
```diff
@@ -153,3 +153,20 @@ export const removeUserFromOrganization = async (
...
+  input: UserOrganizationMutator
+): Promise<UserOrganization> => {
+  const [updatedUserOrganization] = await db<UserOrganization>(
+    context,
+    'User_Organization'
+  )
+    .where(field)
+    .update(input)
+    .returning('*')
+    .secureQuery();
```
Comment (carinelebas) [inline apps/portal-api/src/modules/common/user-organization.domain.ts:169]: do we need the secureQuery? I though we wanted to replace them by another way to check security?

File: apps/portal-api/src/modules/users/users.profile.app.ts [hunk 2]
```diff
@@ -39,4 +48,55 @@ export const usersProfileApp = {
 
     return mappedUser;
   },
+  requestTransferPersonalSpace: async (context, newEmail: string) => {
```
Comment (carinelebas) [inline apps/portal-api/src/modules/users/users.profile.app.ts:51]: no tests for these functions ? 😢

File: apps/portal-api/src/modules/organizations/organizations.domain.ts [hunk 3]
```diff
@@ -29,6 +29,15 @@ export const loadOrganizationsByUser = async (
     .select('Organization.*');
 };
 
+export const loadUserByOrganization = async (
```
Comment (jpkha) [inline apps/portal-api/src/modules/organizations/organizations.domain.ts:32]: I’m not sure if it adds much extra work, but to avoid fetching User.* every time when you only need User.name or User.id, you can call the function like this:

`await loadUserByOrganization().select('User.id')`

This way, we can make the transaction a bit more lightweight.

File: apps/portal-api/src/modules/organizations/organizations.domain.ts [hunk 4]
```diff
@@ -76,22 +85,33 @@ export const insertNewOrganization = (
...
-export const updateOrganization = async (
-  id: OrganizationId,
+export const updateOrganizationBy = async (
+  field: OrganizationMutator,
   data: OrganizationMutator,
   trx?: Knex.Transaction
 ) => {
-  const query = dbUnsecure<Organization>('Organization')
-    .where({ id: id })
+  let query = dbUnsecure<Organization>('Organization')
```
Comment (jpkha) [inline apps/portal-api/src/modules/organizations/organizations.domain.ts:93]: I cannot propose a suggestion directly but what we can do in more elegant way :

```
dbUnsecure<Organization>('Organization')
    .where(field)
    .update(data)
    .returning('*')
    .modify((qb) => {
      if (trx) qb.transacting(trx);
    });```

Comment (jpkha) [reply apps/portal-api/src/modules/organizations/organizations.domain.ts:93]: As I mentioned earlier, you don’t need to force returning('*'). You can let the caller decide which fields to return instead

File: apps/portal-api/src/modules/organizations/organizations.domain.ts [hunk 5]
```diff
@@ -76,22 +85,33 @@ export const insertNewOrganization = (
...
+) => {
+  let query = dbUnsecure<Organization>('Organization')
+    .delete('*')
+    .where(conditions)
+    .returning('*');
+
+  if (trx) {
+    query = query.transacting(trx);
+  }
+  return query;
```
Comment (jpkha) [inline apps/portal-api/src/modules/organizations/organizations.domain.ts:116]:
```suggestion
export const deleteOrganizationBy = (
  conditions: OrganizationMutator,
  trx?: Knex.Transaction
) => {
  return dbUnsecure<Organization>('Organization')
    .where(conditions)
    .delete()
    .modify((qb) => {
      if (trx) qb.transacting(trx);
    })
     .returning('*');
```

File: apps/portal-api/src/modules/subcription/subscription.domain.ts [hunk 6]
```diff
@@ -137,3 +138,19 @@ export const loadSubscriptionBy = async (
...
+  let query = dbUnsecure<Subscription>('Subscription')
+    .where(field)
+    .update(data)
+    .returning('*');
+
+  if (trx) {
+    query = query.transacting(trx);
+  }
+  return query;
+};
```
Comment (jpkha) [inline apps/portal-api/src/modules/subcription/subscription.domain.ts:156]:
```suggestion
export const updateSubscriptionBy = async (
  field: SubscriptionMutator,
  data: SubscriptionMutator,
  trx?: Knex.Transaction
): Promise<Subscription[]> => {
  return dbUnsecure<Subscription>('Subscription')
    .where(field)
    .update(data)
    .returning('*')
    .modify((qb) => {
      if (trx) qb.transacting(trx);
    });
};
```

File: apps/portal-api/src/modules/users/users.profile.app.ts [hunk 7]
```diff
@@ -39,4 +48,55 @@ export const usersProfileApp = {
...
+      to: newEmail,
+      template: 'request_transfer_personal_space',
+      params: {
+        recipientName: `${newUser[0].first_name} ${newUser[0].last_name}`,
+        recipientId: newUser[0].id,
+        previousUserId: context.user.id,
+        previousUserEmail: context.user.email,
+        previousUserName: `${context.user.firstname} ${context.user.lastname}`,
+      },
+    });
```
Comment (jpkha) [inline apps/portal-api/src/modules/users/users.profile.app.ts:73]: If you remove the `select('*')` from the loadUserByOrganization you can do that :

```suggestion
const newUser = await loadUserByOrganization(existingPersonalSpace.id)
await loadUserByOrganization(existingPersonalSpace.id)
  .select('id', 'first_name', 'last_name', 'email')
  .first();
.first();

await sendMail({
  to: newEmail,
  template: 'request_transfer_personal_space',
  params: {
    recipientName: `${newUser.first_name} ${newUser.last_name}`,
    recipientId: newUser.id,
    previousUserId: context.user.id,
    previousUserEmail: context.user.email,
    previousUserName: `${context.user.firstname} ${context.user.lastname}`,
  },
});
```
